src/hooks/useGetPledgeOrDonation.ts

```diff
@@ -0,0 +1,183 @@
+import { useEffect, useState } from 'react';
+import { TFunction } from 'i18next';
+import { DateTime } from 'luxon';
+import { useTranslation } from 'react-i18next';
+import { AppealStatusEnum } from 'src/components/Tool/Appeal/AppealsContext/AppealsContext';
+import { AppealContactInfoFragment } from 'src/components/Tool/Appeal/AppealsContext/contacts.generated';
+import { PledgeFrequencyEnum } from 'src/graphql/types.generated';
+import { currencyFormat, dateFormat } from 'src/lib/intlFormat';
+import { getLocalizedPledgeFrequency } from 'src/utils/functions/getLocalizedPledgeFrequency';
+
+type FormatPledgeOrDonationProps = {
+  amount?: number | null;
+  currency?: string | null;
+  appealStatus: AppealStatusEnum;
+  dateOrFrequency?: PledgeFrequencyEnum | string | null;
+  locale: string;
+  t: TFunction;
+};
+
+const formatPledgeOrDonation = ({
+  amount,
+  currency,
+  appealStatus,
+  dateOrFrequency,
+  locale,
+  t,
+}: FormatPledgeOrDonationProps) => {
+  const pledgeOrDonationAmount =
+    amount && currency
+      ? currencyFormat(amount, currency, locale)
+      : amount || currencyFormat(0, currency, locale);
+
+  let pledgeOverdue = false;
+  if (
+    (appealStatus === AppealStatusEnum.NotReceived ||
+      appealStatus === AppealStatusEnum.ReceivedNotProcessed) &&
+    dateOrFrequency
+  ) {
+    const date = DateTime.fromISO(dateOrFrequency).startOf('day');
+    if (date <= DateTime.local().startOf('day')) {
+      pledgeOverdue = true;
+    }
+  }
+
+  const pledgeOrDonationDate =
+    appealStatus === AppealStatusEnum.Asked ||
+    appealStatus === AppealStatusEnum.Excluded
+      ? (dateOrFrequency &&
+          getLocalizedPledgeFrequency(
+            t,
+            dateOrFrequency as PledgeFrequencyEnum,
+          )) ??
+        ''
+      : dateOrFrequency
+      ? dateFormat(DateTime.fromISO(dateOrFrequency), locale)
+      : null;
+  return {
+    amount: pledgeOrDonationAmount,
+    dateOrFrequency: pledgeOrDonationDate,
+    pledgeOverdue,
+  };
+};
+
+export interface UseGetPledgeOrDonation {
+  pledgeValues: AppealContactInfoFragment['pledges'][0] | undefined;
+  amountAndFrequency: string[] | undefined;
+  pledgeDonations: string[] | null;
+  pledgeOverdue: boolean;
+}
+// The return value doesn't change until `delay` milliseconds have elapsed since the last time `value` changed
+export const useGetPledgeOrDonation = (
+  appealStatus: AppealStatusEnum,
+  contact: AppealContactInfoFragment,
+  appealId: string,
+  locale: string,
+): UseGetPledgeOrDonation => {
+  const { t } = useTranslation();
+  const [pledgeValues, setPledgeValues] =
+    useState<AppealContactInfoFragment['pledges'][0]>();
+  const [amountAndFrequency, setAmountAndFrequency] = useState<string[]>();
+  const [pledgeDonations, setPledgeDonations] = useState<string[] | null>(null);
+  const [pledgeOverdue, setPledgeOverdue] = useState(false);
+
+  useEffect(() => {
```

Comment:
I think that this is another `state/useEffect` combo that could be `useMemo` instead since declarative code is usually easier to reason about than imperative code. Your `useMemo` can return the values for `pledgeValues`, `amountAndFrequency`, `pledgeDonations`, `pledgeOverdue`. It also avoids potential logic bugs since there are some paths through the `useEffect` where not every state variable is being updated.

Reply:
Switched to useMemo. I'm not sure If it made better sense, maybe you could help me understand why people prefer declarative code over imperative code?

Reply:
In my components and hooks, I try to have as few `useState`s as possible and derive any other state that I need from those root states. That way if you want to know the value of a particular sub-state, you just have to look in one place: where it is calculated. On the other hand, if the component/hook uses `useState/useEffect` instead of derived state, I have to look at the `useState` to see its initial value, look at the `useEffect` to see under what conditions it changes and what new values it changes too, and search through the entire component to make sure there aren't any set state calls anywhere else that I need to consider. The derived state version requires me to hold less context in my head when reading the code. As an added benefit, setting state can cause an extra render (https://julvo.com/posts/react/derived-state/).

I'm not sure if I'm explaining it well. Does that make sense? Feel free to push back too.

Reply:
That makes a lot of sense when reviewing code.  Thank you for explaining. 

I did wonder if the useState caused another render.